src/vdaf/idpf.rs

```diff
@@ -0,0 +1,1620 @@
+                "invalid number of bits: 0".to_string(),
+            ));
+        }
+        Ok(IdpfPoplar {
+            bits,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// An auxiliary function that acts as a pseudorandom generator, extending one seed into two
+    /// seeds plus some additional bits.
+    fn extend(&self, seed: &Seed<L>) -> ([Seed<L>; 2], [bool; 2]) {
+        let mut prg = P::init(&seed.0);
+        prg.update(VERSION);
+        prg.update(b" idpf poplar extend");
+        let mut seed_stream = prg.into_seed_stream();
+
+        let mut seeds = [Seed([0u8; L]), Seed([0u8; L])];
+        seed_stream.fill(&mut seeds[0].0);
+        seed_stream.fill(&mut seeds[1].0);
+
+        let mut byte = [0u8];
+        seed_stream.fill(&mut byte);
+        let control_bits = [(byte[0] & 1) != 0, ((byte[0] >> 1) & 1) != 0];
+
+        (seeds, control_bits)
+    }
+
+    /// An auxiliary function that acts as a pseudorandom generator, returning field elements.
+    fn convert<F>(&self, seed: &Seed<L>) -> (Seed<L>, [F; OUT_LEN])
+    where
+        F: FieldElement,
+    {
+        let mut prg = P::init(&seed.0);
+        prg.update(VERSION);
+        prg.update(b" idpf poplar convert");
+        let mut seed_stream = prg.into_seed_stream();
+
+        let mut next_seed = Seed([0u8; L]);
+        seed_stream.fill(&mut next_seed.0);
+
+        let prng = Prng::from_seed_stream(seed_stream);
+        let mut w = [F::zero(); OUT_LEN];
+        for (w_i, output) in w.iter_mut().zip(prng) {
+            *w_i = output;
+        }
+
+        (next_seed, w)
+    }
+
+    /// Helper method to update seeds, update control bits, and output correction words for one
+    /// level of the IDPF key generation process.
+    fn generate_correction_words<F>(
+        &self,
+        input_bit: bool,
+        value: [F; OUT_LEN],
+        seeds: &mut [Seed<L>; 2],
+        control_bits: &mut [bool; 2],
+    ) -> IdpfPoplarCorrectionWords<F, L, OUT_LEN>
+    where
+        F: FieldElement,
+    {
+        // expand both seeds into two more seeds, two more control bits each
+        let (seed_0, control_bits_0) = self.extend(&seeds[0]);
+        let (seed_1, control_bits_1) = self.extend(&seeds[1]);
+
+        let (keep, lose) = (input_bit as usize, !input_bit as usize);
+
+        let seed_correction_word = xor_seeds(&seed_0[lose], &seed_1[lose]);
+        let control_bit_correction_words = [
+            control_bits_0[0] ^ control_bits_1[0] ^ input_bit ^ true,
+            control_bits_0[1] ^ control_bits_1[1] ^ input_bit,
+        ];
+
+        let previous_control_bits = *control_bits;
+        control_bits[0] =
+            control_bits_0[keep] ^ (previous_control_bits[0] & control_bit_correction_words[keep]);
+        control_bits[1] =
+            control_bits_1[keep] ^ (previous_control_bits[1] & control_bit_correction_words[keep]);
+
+        let seeds_corrected = [
+            if previous_control_bits[0] {
```

Comment:
We need to rewrite this code to be constant time, i.e., not branch on the value of the secret control bits. (I'm making a note to adjust the spec accordingly.) One idea is to convert the control bits to bitmasks via wrapping subtraction:

```rust
let control_bit_mask_0 = 0_u8.wrapping_sub(previous_control_bits[0] as u8);
let control_bit_mask_1 = 0_u8.wrapping_sub(previous_control_bits[1] as u8);
for i in 0..L {
    seed_0[keep].0[i] ^= seed_correction_word.0[i] & control_bit_mask_0;
    seed_1[keep].0[i] ^= seed_correction_word.0[i] & control_bit_mask_1;
}
```